### DISCUSSION TITLE
How to get init args & values of a class instance

### DISCUSSION THREAD

**Discussion by Nic-Ma** [Project-MONAI/MONAI#2343] [Ideas]
Hi @ericspod @wyli @vfdev-5 ,

I got an important feature request from the internal team: how to get the init args and the values from a MONAI or ignite class instance and then used for logging or exporting to JSON files, etc.
For example:
```py
from monai.transforms import LoadImaged
load_image = LoadImaged(keys=["image", "seg"], meta_key_postfix="11")
```
How to get `{"keys": ["image", "seg"], "meta_key_postfix": "11"}` from the `load_image` instance?
I tried `inspect`, but seems it can only get the self.XXX instead of the real original init args, here is my test program:
```py
class MONAIClass:
    def __init__(self, keys, count, flag) -> None:
        self.keys = keys
        self.count = count * 10
        print(flag)

load_image = MONAIClass(keys=["image", "seg"], count=1, flag="in class")
import inspect
signature = inspect.signature(load_image.__class__)
print(f"class name: {load_image.__class__}")
print("Arguments")
for key in signature.parameters:
    try:
        value = getattr(load_image, key)
    except Exception as e:
        value = signature.parameters[key].default
    print(key, ":", value)
```
And the the output:
```
class name: <class '__main__.TestAddChanneld.test_class.<locals>.MONAIClass'>
Arguments
keys : ['image', 'seg']
count : 10
flag : <class 'inspect._empty'>
```
**count** and **flag** are not correct here.
Do you python experts have some cool ideas for this problem?
I am blocked by this problem and didn't get a good solution so far, appreciate any help here!

Thanks in advance.

**Comment by vfdev-5** [comment]
Hi @Nic-Ma 

How about inspecting `__dict__`:
```python
from monai.transforms import LoadImaged
load_image = LoadImaged(keys=["image", "seg"], meta_key_postfix="11")

print(load_image.__dict__)
> 
{'keys': ('image', 'seg'), 'allow_missing_keys': False, '_loader': <monai.transforms.io.array.LoadImage object at 0x7f0082845eb0>, 'meta_key_postfix': '11', 'overwriting': False}
```

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

Thanks for your help here.
Sorry, maybe I didn't make the question clear.
The feature request is to get the **args** and the **values** of the "__init__()" func from an instance, not the member properties.

Thanks.

**Comment by vfdev-5** [comment]
Thanks for the explanation! Let me check what we are doing to check the signature...

> I got an important feature request from the internal team: how to get the init args and the values from a MONAI or ignite class instance and then used for logging or exporting to JSON files, etc.

This looks like about logging a configuration which I think this should be done separately:
```
from monai.transforms import LoadImaged

config = {
    "keys": ["image", "seg"], 
    "meta_key_postfix": "11"
}
log_config(config)

load_image = LoadImaged(**config)
```

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

The main purpose of this feature is to export a config from a regular MONAI program which is based on pytorch and ignite.
We don't want to force users to write the config itself, so the ideal solution is to automatically extract args and values from an instance. Users can just call the func to export when they want:
```py
from monai.transforms import LoadImaged
load_image = LoadImaged(keys=["image", "seg"], meta_key_postfix="11")

net = ...
loss = ...
trainer =...
trainer.run()

# call export to get the config and save into JSON
export_config([trainer, net, loss, load_image])
```

Thanks.

**Reply by vfdev-5** [reply]
Thanks for the details. Object's introspection probably can help as well. 

If the purpose is to enable code reproducibility, a best practice is to log the commit and all uncommitted changes to the complete code. Such that if user forgets to include some other objects to `export_config` function, everything is "recorded" by the commit and uncommitted changes...

**Reply by Nic-Ma** [reply]
Hi @vfdev-5 , thanks for your suggestions!
The main purpose is to automatically generate the MMAR JSON config of NVIDIA Clara from a MONAI program.

Thanks.

**Comment by wyli** [comment]
perhaps you can trace the call like this one: https://stackoverflow.com/a/8315566 and filter the relevant constructor calls. (I'm converting the ticket to a discussion)

**Reply by yiheng-wang-nv** [reply]
> perhaps you can trace the call like this one: https://stackoverflow.com/a/8315566 and filter the relevant constructor calls. (I'm converting the ticket to a discussion)

Thanks @wyli , let me look at this line.

**Comment by ericspod** [comment]
Another way would be to monkey-patch the constructor to capture arguments somewhere when called, using `inspect` to identify what arguments are given. This would be a bit messier as you'd have to identify which classes to apply this to and have code traverse the whole module tree to find what classes to patch.

**Reply by Nic-Ma** [reply]
Thanks @wyli @ericspod for all your suggestions, let Yiheng and I have a try and get back to you soon.
